pkg/ipam-controller/allocator/allocator.go

```diff
@@ -0,0 +1,373 @@
+	log := pa.getLog(ctx, cfg)
+	log.V(1).Info("pool configuration update")
+	if pa.cfg.Equal(&cfg) {
+		log.V(1).Info("pool configuration is the same, keep allocations")
+		return
+	}
+	pa.cfg = cfg
+	pa.allocations = map[string]allocatedRange{}
+	log.Info("pool configuration updated, reset allocations")
+}
+
+// Allocate allocates a new range in the poolAllocator or
+// return existing one,
+// returns ErrNoFreeRanges if no free ranges available
+func (pa *poolAllocator) Allocate(ctx context.Context, node string) (nodeAllocationInfo, error) {
+	log := pa.getLog(ctx, pa.cfg).WithValues("node", node)
+	existingAlloc, exist := pa.allocations[node]
+	if exist {
+		log.V(1).Info("allocation for the node already exist",
+			"start", existingAlloc.StartIP, "end", existingAlloc.EndIP)
+		return pa.getNodeAllocationInfo(node, existingAlloc), nil
+	}
+	allocations := pa.getAllocationsAsSlice()
+	var startIP net.IP
+	if len(allocations) == 0 || ip.Distance(ip.Network(pa.cfg.Subnet).IP, allocations[0].StartIP) > 2 {
+		// start allocations from the network address if there are no allocations or if the "hole" exist before
+		// the firs allocation
+		startIP = ip.NextIP(ip.Network(pa.cfg.Subnet).IP)
+	} else {
+		for i := 0; i < len(allocations); i++ {
+			nextI := i + 1
+			// if last allocation in the list
+			if nextI > len(allocations)-1 ||
+				// or we found a "hole" in allocations. the "hole" can't be less than required for the allocation by design.
+				// because we reset all allocations when PerNodeBlockSize size changes
+				ip.Distance(allocations[i].EndIP, allocations[nextI].StartIP) > 1 {
+				startIP = ip.NextIP(allocations[i].EndIP)
+				break
+			}
+		}
+	}
+	nodeRange := allocatedRange{
+		StartIP: startIP,
+		EndIP:   ip.NextIPWithOffset(startIP, int64(pa.cfg.PerNodeBlockSize)-1),
+	}
+	if !pa.cfg.Subnet.Contains(nodeRange.EndIP) || ip.IsBroadcast(nodeRange.EndIP, pa.cfg.Subnet) {
+		// out of range
+		log.Info("can't allocate: pool has no free ranges")
+		return nodeAllocationInfo{}, ErrNoFreeRanges
+	}
+
+	pa.allocations[node] = nodeRange
+	log.Info("range allocated",
+		"start", nodeRange.StartIP, "end", nodeRange.EndIP)
+	return pa.getNodeAllocationInfo(node, nodeRange), nil
+}
+
+// Deallocate remove info about allocation for the node from the poolAllocator
+func (pa *poolAllocator) Deallocate(ctx context.Context, node string) {
+	log := pa.getLog(ctx, pa.cfg)
+	log.Info("deallocate range for node", "node", node)
+	delete(pa.allocations, node)
+}
+
+// Load loads range to the pool allocator with validation for conflicts
+func (pa *poolAllocator) Load(ctx context.Context, allocData nodeAllocationInfo) error {
+	log := pa.getLog(ctx, pa.cfg)
+	if err := pa.checkAllocation(allocData); err != nil {
+		log.Info("range check failed", "reason", err.Error())
+		return err
+	}
+	allocations := pa.getAllocationsAsSlice()
+	for _, a := range allocations {
+		if allocData.allocatedRange.StartIP.Equal(a.StartIP) {
```

Comment:
can you add a comment here on why this is enough ? 

to the naive reader it many not be apparent.